src/util/skiplist.c

```diff
@@ -0,0 +1,451 @@
+    update[i]->level[i].span++;
+  }
+
+  x->backward = (update[0] == sl->header) ? NULL : update[0];
+  if (x->level[0].forward) {
+    x->level[0].forward->backward = x;
+  } else {
+    sl->tail = x;
+  }
+  sl->length++;
+
+  return x;
+}
+
+/*
+ * Internal function used by skiplistDelete, it needs an array of other
+ * skiplist nodes that point to the node to delete in order to update
+ * all the references of the node we are going to remove.
+ */
+void skiplistDeleteNode(skiplist *sl, skiplistNode *x, skiplistNode **update) {
+  // Free the node key if we specified a routine for it
+  if (sl->freeObj) sl->freeObj(x->obj);
+
+  int i;
+  for (i = 0; i < sl->level; i++) {
+    if (update[i]->level[i].forward == x) {
+      update[i]->level[i].span += x->level[i].span - 1;
+      update[i]->level[i].forward = x->level[i].forward;
+    } else {
+      update[i]->level[i].span -= 1;
+    }
+  }
+  if (x->level[0].forward) {
+    x->level[0].forward->backward = x->backward;
+  } else {
+    sl->tail = x->backward;
+  }
+  while (sl->level > 1 && sl->header->level[sl->level - 1].forward == NULL) {
+    sl->level--;
+  }
+  sl->length--;
+}
+
+/*
+ * Delete an element from the skiplist. If the `val` argument is provided,
+ * the single matching value will be found and deleted; otherwise, the skiplistNode
+ * representing the key matching `obj` and all of its elements will be deleted.
+ * If the operation is successful, 1 is returned, otherwise the target was not found and 0 is returned.
+ * The memory used to to represent skiplist keys and values is not freed.
+ */
+int skiplistDelete(skiplist *sl, void *obj, void *val) {
+  skiplistNode *update[SKIPLIST_MAXLEVEL], *x;
+  int i;
+
+  x = sl->header;
+  for (i = sl->level - 1; i >= 0; i--) {
+    while (x->level[i].forward &&
+           sl->compare(x->level[i].forward->obj, obj, sl->cmpCtx) < 0) {
+      x = x->level[i].forward;
+    }
+    update[i] = x;
+  }
+  x = x->level[0].forward;
+  if (x && sl->compare(x->obj, obj, sl->cmpCtx) == 0) {
+
+    if (val) {
+      // try to delete the value itself from the vallist
+      int found = 0;
+      for (i = 0; i < x->numVals; i++) {
+        // found the value - let's delete it
+        if (!sl->valcmp(val, x->vals[i])) {
+          found = 1;
+          // switch the found value with the top value
+          if (i < x->numVals - 1) {
+            x->vals[i] = x->vals[x->numVals - 1];
+          }
+          x->numVals--;
+          break;
+        }
+      }
+      if (!found) {
+        // Specified value was not found in skiplistNode
+        return 0;
+      }
+    }
+
+    if (!val || x->numVals == 0) {
+      skiplistDeleteNode(sl, x, update);
+      skiplistFreeNode(x);
+    }
+    return 1;
+  }
+  return 0; /* not found */
+}
+
+/*
+ * Search for a specific element among values associated to a single skiplist key
+ */
+void *searchSkiplistNode(skiplistNode *sl_node, void *value, skiplistValCmpFunc valcmp) {
```

Comment:
Why do we need this function, I can only see it's being used in the test suite.
Also I'm not sure about the return value, maybe it would be better to return the position in which the value is located at if found and -1 if the value isn't in the array.